pkg/mgmt/snapshot/snapshot.go

```diff
@@ -0,0 +1,249 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package snapshot
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/Sirupsen/logrus"
+
+	apis "github.com/openebs/zfs-localpv/pkg/apis/openebs.io/core/v1alpha1"
+	zfs "github.com/openebs/zfs-localpv/pkg/zfs"
+	k8serror "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/tools/cache"
+)
+
+// isDeletionCandidate checks if a zfs snapshot is a deletion candidate.
+func (c *SnapController) isDeletionCandidate(snap *apis.ZFSSnapshot) bool {
+	return snap.ObjectMeta.DeletionTimestamp != nil
+}
+
+// syncHandler compares the actual state with the desired, and attempts to
+// converge the two. It then updates the Status block of the spcPoolUpdated resource
+// with the current status of the resource.
+func (c *SnapController) syncHandler(key string) error {
+	// Convert the namespace/name string into a distinct namespace and name
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		runtime.HandleError(fmt.Errorf("invalid resource key: %s", key))
+		return nil
+	}
+
+	// Get the snap resource with this namespace/name
+	snap, err := c.snapLister.ZFSSnapshots(namespace).Get(name)
+	if k8serror.IsNotFound(err) {
+		runtime.HandleError(fmt.Errorf("zfs snapshot '%s' has been deleted", key))
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+	snapCopy := snap.DeepCopy()
+	err = c.syncSnap(snapCopy)
+	return err
+}
+
+// enqueueSnap takes a ZFSSnapshot resource and converts it into a namespace/name
+// string which is then put onto the work queue. This method should *not* be
+// passed resources of any type other than ZFSSnapshot.
+func (c *SnapController) enqueueSnap(obj interface{}) {
+	var key string
+	var err error
+	if key, err = cache.MetaNamespaceKeyFunc(obj); err != nil {
+		runtime.HandleError(err)
+		return
+	}
+	c.workqueue.Add(key)
+}
+
+// synSnap is the function which tries to converge to a desired state for the
+// ZFSSnapshot
+func (c *SnapController) syncSnap(snap *apis.ZFSSnapshot) error {
+	var err error
+	// ZFSSnapshot should be deleted. Check if deletion timestamp is set
+	if c.isDeletionCandidate(snap) {
+		err = zfs.DestroySnapshot(snap)
+		if err == nil {
+			zfs.RemoveSnapFinalizer(snap)
+		}
+	} else {
+		// if finalizer is not set then it means we are creating
+		// the zfs snapshot.
+		if snap.Finalizers == nil {
+			err = zfs.CreateSnapshot(snap)
+			if err == nil {
+				err = zfs.UpdateSnapInfo(snap)
+			}
+		}
```

Comment:
Consider a case when snapshot creation is successful and snapshot status fails to update a ready state. 
I am suspecting the state may not recover to the `ready` state.

Reply:
it will reconcile again.